File: MinMax/MinMax.js
/*
La seguente funzione trova il valore della mossa migliore usando l'algoritmo di MinMax senza alpha beta pruning
Parametro -> T = L'albero delle mosse contente il valore della mossa ed i figli
Parametro -> depth = Profondita con cui si vuole cercare nell'albero
Parametro ->  maximizingPlayer = Indica se il giocatore corrente cerca la mossa dal valore più alto (true) o più basso (false)
Return ->  il valore della mossa migliore
*/
function MinMax(T, depth, maximizingPlayer) {
    if(depth == 0)
        return T.value;
    let best_Value, v
    if(maximizingPlayer){
        best_Value = Number.NEGATIVE_INFINITY
        for(let i = 0; i<T.childs.length; i++){
            v = MinMax(T.childs[i], depth - 1, false)
            best_Value = Math.max(best_Value, v)
        }
        return best_Value
    }
    else{
        best_Value = Number.POSITIVE_INFINITY
        for(let i = 0; i<T.childs.length; i++){
            v = MinMax(T.childs[i], depth - 1, true)
            best_Value = Math.min(best_Value, v)
        }
        return best_Value
    }
}

let tree = {value:0,
    childs:[
        {value:3,
            childs:[
                {value:10, childs:[
                    {value: 9, childs:[]},
                    {value: 1, childs:[]}
                ]}, 
                {value:2, childs: [
                    {value:-5, childs:[]},
                    {value:-3, childs:[]}
                ]}]
        }, 
        {value:5,
            childs:[
                {value:4, childs:[
                    {value:33, childs:[]},
                    {value:-10, childs:[]}
                ]}, 
                {value:7, childs:[
                    {value:42, childs:[]},
                    {value:30, childs:[]}
                ]}]
        }
    ]
}
console.log(MinMax(tree, 3, true))

File: MinMax/MinMax.py
import math

class Tree:
    def __init__(self, data):
        self.data = data
        self.childs: list[Tree] = []


def MinMax(T: Tree, depth, isMaximizingPlayer):
    if depth == 0:
        return T.data
    if isMaximizingPlayer:
        best_value = -math.inf
        for child in T.childs:
            eval = MinMax(child, depth-1, False)
            best_value = max(best_value, eval)
        return best_value
    else:
        best_value = math.inf
        for child in T.childs:
            eval = MinMax(child, depth-1, True)
            best_value = min(best_value, eval)
        return best_value


root = Tree(0)
root.childs.append(Tree(0))
root.childs.append(Tree(0))
root.childs[0].childs.append(Tree(1))
root.childs[0].childs.append(Tree(-1))
root.childs[1].childs.append(Tree(0))
root.childs[1].childs.append(Tree(-1))
print(MinMax(root, 1, True))